wrappers/kn/src/nativeMain/kotlin/zxingcpp/Barcode.kt

```diff
@@ -0,0 +1,117 @@
+package zxingcpp
+
+import cnames.structs.zxing_Barcode
+import cnames.structs.zxing_Barcodes
+import kotlinx.cinterop.*
+import zxingcpp.cinterop.*
+import zxingcpp.cinterop.zxing_ContentType.*
+import kotlin.experimental.ExperimentalNativeApi
+import kotlin.native.ref.createCleaner
+
+@OptIn(ExperimentalForeignApi::class)
+enum class ContentType(internal val cValue: zxing_ContentType) {
+	Text(zxing_ContentType_Text),
+	Binary(zxing_ContentType_Binary),
+	Mixed(zxing_ContentType_Mixed),
+	GS1(zxing_ContentType_GS1),
+	ISO15434(zxing_ContentType_ISO15434),
+	UnknownECI(zxing_ContentType_UnknownECI);
+}
+
+@OptIn(ExperimentalForeignApi::class)
+fun zxing_ContentType.toKObject(): ContentType {
+	return ContentType.entries.first { it.cValue == this }
+}
+
+data class PointI(
+	val x: Int,
+	val y: Int
+)
+
+@OptIn(ExperimentalForeignApi::class)
+fun zxing_PointI.toKObject(): PointI = PointI(x, y)
+
+data class Position(
+	val topLeft: PointI,
+	val topRight: PointI,
+	val bottomRight: PointI,
+	val bottomLeft: PointI,
+)
+
+@OptIn(ExperimentalForeignApi::class)
+fun zxing_Position.toKObject(): Position = Position(
+	topLeft.toKObject(),
+	topRight.toKObject(),
+	bottomRight.toKObject(),
+	bottomLeft.toKObject(),
+)
+
+@OptIn(ExperimentalForeignApi::class)
+class Barcode(val cValue: CValuesRef<zxing_Barcode>) {
+	val isValid: Boolean
+		get() = zxing_Barcode_isValid(cValue)
+	val errorMsg: String?
+		get() = zxing_Barcode_errorMsg(cValue)?.toKStringAndFree()
+	val format: BarcodeFormat
+		get() = zxing_Barcode_format(cValue).parseIntoBarcodeFormat().first { it != BarcodeFormat.None }
+	val contentType: ContentType
+		get() = zxing_Barcode_contentType(cValue).toKObject()
+
+	fun bytes(len: Int): ByteArray? = zxing_Barcode_bytes(cValue, cValuesOf(len))?.run {
+		readBytes(len).also { zxing_free(this) }
+	}
+
+	fun bytesECI(len: Int): ByteArray? = zxing_Barcode_bytesECI(cValue, cValuesOf(len))?.run {
+		readBytes(len).also { zxing_free(this) }
+	}
+
+	val text: String?
+		get() = zxing_Barcode_text(cValue)?.toKStringAndFree()
+	val ecLevel: String?
+		get() = zxing_Barcode_ecLevel(cValue)?.toKStringAndFree()
+	val symbologyIdentifier: String?
+		get() = zxing_Barcode_symbologyIdentifier(cValue)?.toKStringAndFree()
+	val position: Position
+		get() = zxing_Barcode_position(cValue).useContents { toKObject() }
+	val orientation: Int
+		get() = zxing_Barcode_orientation(cValue)
+	val hasECI: Boolean
+		get() = zxing_Barcode_hasECI(cValue)
+	val isInverted: Boolean
+		get() = zxing_Barcode_isInverted(cValue)
+	val isMirrored: Boolean
+		get() = zxing_Barcode_isMirrored(cValue)
+	val lineCount: Int
+		get() = zxing_Barcode_lineCount(cValue)
+
+	@Suppress("unused")
+	@OptIn(ExperimentalNativeApi::class)
+	private val cleaner = createCleaner(cValue) { zxing_Barcode_delete(it) }
+
+	override fun toString(): String {
```

Comment:
This is just a helper function for typical kotlin value class, which can easily generate a human readable string.
If $text returns null, then it will result in `text=null`, and no exception will be generated.

> This is related to the question why all the String properties are nullable?
For this, that's because kotlin cannot predict whether the c-side function will return `NULL` pointer, so it treat all of them as `nullable`. I can add an null check here if you want it.

Reply:
In kotlin, `null` and `String("")` has completely different meaning.
In my opinon:
1. the data doesn't exist or not applicable -> `null`
2. empty data -> `String("")`
3. things should never be null (e.g. type of something) -> throw exception in kotlin (and remove `?`)

Reply:
Actually, for the case you mentioned, I think throw out a exception is better because access to the text of an invalid qrcode is meaningless?

> Your 3 points seem contradictory to me.

I think the 3 points should be arranged as follow, which maybe less confusing (Maybe I should put more time onto learning english after this winter vacation 😂) (This is what I follow when developing kotlin)
1. things that can be "not exist" and accessing to it in this case makes sense
a) the data doesn't exist -> `null`
b) empty data -> `String("")`
c) real data-> `String`
d) something went wrong generating the value -> throw exception
2. things should never be null (e.g. type of something)
a) real data-> `String`
b) something went wrong generating the value -> throw exception

Reply:
When it comes to binding, the exception cannot be correctly handled by kotlin(At least I don't know howto, maybe there's no such thing like checked exception in pure-c?)
So case d in 1 and b in 2 have to be mapped to null or String(""). I think we are facing this because there's no corresponding concept for exception.
So maybe we will **have to** "give up" a case in 1(a or d)? (I mean we abandon something to obtain something, how to express this precisely in english?) a case in 1(a or d)?

I comment these because I don't completely understand why do you map like that (Seems you choose to "give up" b in 1)

Reply:
Well, since the C side only returns `NULL` if `malloc` fails, I see that as an exception too.

In any other case, `text` will be a string, maybe empty, and I think any wrapper should work with that, because otherwise the wrapper would reinterpret what an empty string means, which is probably not a good idea.

So for me the correct thing to do would be that the corresponding Kotlin function always returns a `String` object, never `null`, and throws an (OutOfMemory or Runtime) exception if the C side returns `NULL`.

But - this is _not_ what the current Android wrapper is doing 😬  [Here](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/src/main/cpp/ZXingCpp.cpp#L222), we use `Result.isValid` to either return `null` or a `String` object, and [Result.text](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/src/main/java/zxingcpp/BarcodeReader.kt#L106) is of type `String?`. The reason for this was that the JVM's `Result` object does not contain an `isValid` field (at the moment).

So it probaply depends on if this wrapper has an `isValid` in its `Result`/`Barcode` object or not.

Reply:
First of all: you use statements like
```kotlin
?: throw BarcodeReadingException(ZXing_LastErrorMsg()?.toKStringAndFree()
```
so there is a concept of exceptions in Kotlin, right? ;)

Your listing in 1. is completely correct and applies e.g. to the `text` property. And I do indeed suggest to "give up" 1.b) and use 1.a) instead because it is more expressive in Kotlin as it clearly states "there is no content here, something is wrong, please look at `errorMsg`" rather than "the barcode is valid and contains an 'empty' string", which can not happen. You can think of the following class invariant: `isValid == (text != null)`.

Reply:
Thx for your explaining
what's more, does this also applies to `ecLevel`?

> * `NULL` from c -> throw exception
> * empty string from c -> `null`
> * `String` otherwise.

Reply:
@markusfisch I only saw your reply after my last comment. There is an `isValid` property in the c-api (and hence here) because checking `strlen(ZXing_Barcode_text()) > 0` (thereby leaking memory) to get information would be simply horrible from a usability perspective. You are right that my suggestion and what currently is used in the Android wrapper means a conceptual difference between the C/C++ API and Kotlin. But as I argued above: I view this as a better and more idiomatic API in Kotlin compared to C++. Today I might have modeled that as `std::optional<std::string> text()` in c++ but in c++14 there was no such thing. And `std::optional` is still clumsy compared to `String?`.

You could make a point that we don't need `isValid` if we have `text: String?` but the former is even more expressive in my opinion.

Reply:
Agreed, I would prefer a dedicated `isValid` member too.

Reply:
Is this okay?:
```
val errorMsg: String?
	get() = ZXing_Barcode_errorMsg(cValue)?.toKStringAndFree()
val text: String?
	get() = (ZXing_Barcode_text(cValue)?.toKStringAndFree() ?: throw OutOfMemoryError()).ifEmpty { null }
val ecLevel: String?
	get() = (ZXing_Barcode_ecLevel(cValue)?.toKStringAndFree() ?: throw OutOfMemoryError()).ifEmpty { null }
```
or this?:
```
val errorMsg: String?
	get() = (ZXing_Barcode_errorMsg(cValue)?.toKStringAndFree() ?: throw OutOfMemoryError()).ifEmpty { null }
val text: String?
	get() = (ZXing_Barcode_text(cValue)?.toKStringAndFree() ?: throw OutOfMemoryError()).ifEmpty { null }
val ecLevel: String?
	get() = (ZXing_Barcode_ecLevel(cValue)?.toKStringAndFree() ?: throw OutOfMemoryError()).ifEmpty { null }
```

Reply:
@ISNing the second. In fact *every* function returning a `char*` or a `uint8_t*` may return `NULL` if the `malloc` fails. Well, not quite: `ZXing_LastErrorMsg()` is explicitly defined to return `NULL` if there is no last error. I just added a comment to the header file. Anyway, the `bytes() / bytesECI()` members should behave the same as the String members.

Reply:
https://github.com/zxing-cpp/zxing-cpp/pull/719/commits/f348078f0996ad973a0052cc3bbf7ae52475c606 I've commited it and I added a check for `isValid` to avoid throwing OOMError when `isValid==true` which doesn't makes sense.

Reply:
note: `errorMsg` has to behave exactly like all other `Barcode` properties. I was referring to `ZXing_LastErrorMsg` which serves a completely different purpose. Please move the "throw + ifEmpty" handling into the `toKStringAndFree()` helper (maybe give it a different name).

Reply:
fixed: https://github.com/zxing-cpp/zxing-cpp/pull/719/commits/35f985c50524f3aba9438d325d8254065bee4ba6

> note: `errorMsg` has to behave exactly like all other `Barcode` properties. I was referring to `ZXing_LastErrorMsg` which serves a completely different purpose. Please move the "throw + ifEmpty" handling into the `toKStringAndFree()` helper (maybe give it a different name).